ui/app/models/ldap/role.js

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import Model, { attr } from '@ember-data/model';
+import { tracked } from '@glimmer/tracking';
+import { withFormFields } from 'vault/decorators/model-form-fields';
+import { withModelValidations } from 'vault/decorators/model-validations';
+
+const creationLdifExample = `# The example below is treated as a comment and will not be submitted
+# dn: cn={{.Username}},ou=users,dc=learn,dc=example
+# objectClass: person
+# objectClass: top
+`;
+const deletionLdifExample = `# The example below is treated as a comment and will not be submitted
+# dn: cn={{.Username}},ou=users,dc=learn,dc=example
+# changetype: delete
+`;
+const rollbackLdifExample = `# The example below is treated as a comment and will not be submitted
+# dn: cn={{.Username}},ou=users,dc=learn,dc=example
+# changetype: delete
+`;
+
+const validations = {
+  username: [
+    {
+      validator: (model) => (model.isStatic && !model.username ? false : true),
+      message: 'Username is required.',
+    },
+  ],
+  rotation_period: [
+    {
+      validator: (model) => (model.isStatic && !model.rotation_period ? false : true),
+      message: 'Rotation Period is required.',
+    },
+  ],
+  creation_ldif: [
+    {
+      validator: (model) => (model.isDynamic && !model.creation_ldif ? false : true),
+      message: 'Creation LDIF is required.',
+    },
+  ],
+};
+
+export const staticRoleFields = ['username', 'dn', 'rotation_period'];
+export const dynamicRoleFields = [
+  'default_ttl',
+  'max_ttl',
+  'username_template',
+  'creation_ldif',
+  'deletion_ldif',
+  'rollback_ldif',
+];
+
+@withModelValidations(validations)
+@withFormFields()
+export default class LdapRoleModel extends Model {
+  @tracked type; // this must be set to either static or dynamic in order for the adapter to build the correct url
```

Comment:
Why are we making this a tracked value rather than an attribute? I don't think I've seen that in our models before

Reply:
Also, since `fieldsForType` defaults to dynamic, should we make that explicit here?

Reply:
I thought about checking if `type === dynamic` in `fieldsForType` but I'm not sure it will matter since the user has to select a card value in order for the form fields to show up. I could add it though all the same and return an empty array which may help during development.

Reply:
In regards to type as a `tracked` property, it looks like I will need to use `attr` instead since it needs to be set on the records returned from `queryRecord` and I will then have to remove it from the payload in the serializer.